### PR TITLE
MAILBOX-381 Implement Search models for the Vault

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java [hunk 1]
```diff
@@ -141,7 +141,7 @@ public FinalStage(MessageId messageId, List<MailboxId> originMailboxes, User own
             }
 
             public FinalStage subject(String subject) {
-                this.subject = Optional.of(subject);
+                this.subject = Optional.ofNullable(subject);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: Why is this change needed?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    static ListMatcherFactory<MailAddress> recipients() {
+        return () -> RECIPIENTS;
+    }
+
+    static EqualsMatcherFactory<MailAddress> sender() {
+        return () -> SENDER;
+    }
+
+    static EqualsMatcherFactory<Boolean> hasAttachment() {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: I would prefer in this case having 3 methods:

```
static Criterion<T> hasAttachment(boolean expectedValue);
static Criterion<T> hasAttachment();
static Criterion<T> hasNoAttachment();
```

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+            assertThat(search(Query.of(CriterionFactory.sender().equalsMatcher(SENDER3))))
+                .isEmpty();
+        }
+
+        @Test
+        default void shouldNotReturnMessagesWithNullSenderWhenEquals() {
+            DeletedMessage message1 = storeMessageWithSender(SENDER);
+            storeMessageWithSender(SENDER2);
+            storeMessageWithSender(null);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]: `null` spotted. Please use `MaybeSender.nullSender()`

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+
+            assertThat(search(Query.of(
+                    CriterionFactory.subject().containsIgnoreCase("project"))))
+                .containsOnly(message1);
+        }
+
+        @Test
+        default void shouldNotReturnMissingSubjectMessagesWhenContainsIgnoreCase() {
+            DeletedMessage message1 = storeMessageWithSubject(APACHE_JAMES_PROJECT);
+            DeletedMessage message2 = storeMessageWithSubject(null);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]: `null` is not the right way to represent something missing.

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+            .sender(MaybeSender.of(SENDER))
+            .recipients(RECIPIENT1)
+            .content(() -> new ByteArrayInputStream(CONTENT))
+            .hasAttachment(false)
+            .subject(subject)
+            .build();
+
+        Mono.from(getVault().append(USER, deletedMessage))
+            .block();
+        return deletedMessage;
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]:
```
        Mono.from(getVault().append(USER, deletedMessage))
            .block();
        return deletedMessage;
```

is duplicated 8 times. Time for extraction?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java (hunk 1 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: there is no guarantee that the caller will pass a non null subject to `subject(String subject)` method. In that case it throws an unexpected error.

Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: I would argue that it throws an expected `NullPointerException` that is just not yet tested and all of this "fails early" classical tactic prevents later unexpected behaviour.

Let's be "null free" whereever we can. New APIs definitely are such places.

Maybe what you truely want is to pass the `subject` as an optional like I did here: https://github.com/linagora/james-project/pull/2194/commits/ad0550fe869ff80a4bbbd5fd3a15c47594a83fe7

Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: okay, I revered to use the original builder method.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/DeletedMessageField.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.vault.DeletedMessage;
+
+public interface DeletedMessageField<T> {
+    interface ValueExtractor<T> {
+        Optional<T> extract(DeletedMessage deletedMessage);
+
+        ValueExtractor<ZonedDateTime> DELETION_DATE_EXTRACTOR = deletedMessage -> Optional.ofNullable(deletedMessage.getDeletionDate());
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/DeletedMessageField.java]: why should we handle null case ? Does a deleted message without a deletion date exist at all?
Same question for other fields.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    interface Equals<T> extends SingleValueMatcher<T> {
+        @Override
+        default boolean matches(T referenceValue) {
+            return referenceValue.equals(testedValue());
+        }
+    }
+
+    interface StringContains extends SingleValueMatcher<String> {
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: I don't agree about your use of interfaces, it makes things more complex than needed.
If you want to pass a function to a class, just do that but don't overuse interface inheritance.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.List;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailbox.model.MailboxId;
+
+public interface CriterionFactory<T> {
+
+    DeletedMessageField<T> deletedMessageField();
+
+    interface EqualsMatcherFactory<T> extends CriterionFactory<T> {
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: see my comment below : you probably abused interfaces too much

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/DeletedMessageField.java (hunk 6 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/DeletedMessageField.java]: DeletionDate is an compulsory field, base on the DeletedMessage model, Subject or Sender could be missing. then I propose an Optional returned for field extractors. Otherwise I cannot set a default value to subject or sender because they are actually missing

Comment (mbaechler) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/DeletedMessageField.java]: ok but don't use `ofNullable`, use `of` instead

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.time.ZonedDateTime;
+import java.util.List;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailbox.model.MailboxId;
+
+public interface CriterionFactory {
+
+    class StringMatcherFactory {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: This factory is producing `Criterion` and not `Matcher` (though you could argue a criterion is some kind of matcher :trollface:  )

Let me propose to rename it `StringCriterionFactory`

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+        public Criterion<String> contains(String subString) {
+            return new Criterion<>(deletedMessageField, ValueMatcher.SingleValueMatcher.contains(subString));
+        }
+
+        public Criterion<String> containsIgnoreCase(String subString) {
+            return new Criterion<>(deletedMessageField, ValueMatcher.SingleValueMatcher.containsIgnoreCase(subString));
+        }
+
+        public Criterion<String> equals(String testedString) {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: Do we miss a `equalsIgnoreCase` method??

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+            return new Criterion<>(deletedMessageField, ValueMatcher.SingleValueMatcher.isEquals(testedString));
+        }
+    }
+
+    class ZonedDateTimeMatcherFactory {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: DateCriterionFactory

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+public interface ValueMatcher<T> {
+
+    boolean matches(T referenceValue);
+
+    class SingleValueMatcher<T> implements ValueMatcher<T> {
+
+        private static final BiFunction<String, String, Boolean> CONTAINS = (testedValue, referenceValue) -> referenceValue.contains(testedValue);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: I really wonder if all of these class could not be some simple lambdas in Criterion factory...

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    static Criterion<List<MailAddress>> containsRecipient(MailAddress recipient) {
+        return new Criterion<>(RECIPIENTS, ValueMatcher.ListContains.of(recipient));
+    }
+
+    static Criterion<MailAddress> hasSender(MailAddress sender) {
+        return new Criterion<>(SENDER, ValueMatcher.SingleValueMatcher.isEquals(sender));
+    }
+
+    static Criterion<Boolean> hasAttachment() {
+        return new Criterion<>(HAS_ATTACHMENT, ValueMatcher.SingleValueMatcher.isEquals(true));
```
Comment (Arsnael) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: Not better to return then `return hasAttachment(true);` ?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    static Criterion<Boolean> hasNoAttachment() {
+        return new Criterion<>(HAS_ATTACHMENT, ValueMatcher.SingleValueMatcher.isEquals(false));
```
Comment (Arsnael) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: Idem, `return hasAttachment(false);` ?

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+            DeletedMessage message2 = storeDefaultMessage();
+            DeletedMessage message3 = storeDefaultMessage();
+            DeletedMessage message4 = storeMessageWithSender(MaybeSender.nullSender());
+
+            assertThat(search(Query.ALL))
+                .containsOnly(message1, message2, message3, message4);
+        }
+
+        @Test
+        default void searchShouldReturnMessageWhenHavingSameCreterionTypes() {
```
Comment (Arsnael) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]: Typo mistake -> `s/Creterion/Criterion`

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+            DeletedMessage message3 = storeMessageWithRecipients(RECIPIENT1);
+
+            assertThat(search(Query.of(
+                    CriterionFactory.containsRecipient(RECIPIENT1),
+                    CriterionFactory.containsRecipient(RECIPIENT2))))
+                .containsOnly(message1, message2);
+        }
+
+        @Test
+        default void searchShouldReturnEmptyWhenHavingSameCreterionTypesButOppositeMatching() {
```
Comment (Arsnael) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]: Idem

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java (hunk 7 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: I removed multiple interfaces by classes and static methods.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/Query.java [hunk 17]
```diff
@@ -36,7 +39,9 @@ private Query(List<Criterion> criteria) {
         this.criteria = criteria;
     }
 
-    public List<Criterion> getCriteria() {
-        return criteria;
+    public Predicate<DeletedMessage> toPredicate() {
+        return criteria.stream()
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/Query.java]:
```suggestion
        return criteria.stream()
            .map(Criterion::toPredicate)
            .reduce(Predicate::and)
            .orElse(MATCH_ALL);
```

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: Could we find a name for `BiFunction<T, T, Boolean>` ?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    class StringCriterionFactory {
+
+        private final DeletedMessageField<String> deletedMessageField;
+
+        private StringCriterionFactory(DeletedMessageField<String> deletedMessageField) {
+            this.deletedMessageField = deletedMessageField;
+        }
+
+        public Criterion<String> contains(String subString) {
+            return new Criterion<>(deletedMessageField, ValueMatcher.SingleValueMatcher.contains(subString));
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/CriterionFactory.java]: The fact that you reference ValueMatcher into CriterionFactory and CriterionFactory in ValueMatcher should trigger a warning: the separation of concern is probably not good.
BTW, I have hard time figuring out the design you implemented, it should be possible to simplify the design again.

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        private static final BiFunction<String, String, Boolean> CONTAINS_IGNORE_CASE =
+            (testedValue, referenceValue) -> referenceValue.toLowerCase(Locale.US).contains(testedValue.toLowerCase(Locale.US));
```
Comment (mbaechler) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: You can't choose an arbitrary Locale when you match user data because you don't know yet the right Locale you have to pick.

Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: Well, we clearly don't have better solution here.

Maybe the locale of the system would be slightly better.

Do you have better suggestion @mbaechler  ?

Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: Okay, the final decision is keeping this usage of locale and I will document the warning about locale in webadmin markdown in latter PR of Vault API.

Comment (mbaechler) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: > Okay, the final decision is keeping this usage of locale and I will document the warning about locale in webadmin markdown in latter PR of Vault API.

It's not what I suggested (and agreed): I asked to write a failing test, open a JIRA for documentation and disable that test with the JIRA number.
The webadmin documentation is nice too.

**Comment by mbaechler** [comment]
I'd like my commit to be squashed with others to avoid weird history. WDYT?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java (hunk 20 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java]: it is arrived: 99fb4fd

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java [hunk 21]
```diff
@@ -521,6 +523,23 @@ default void searchForAnUserShouldNotReturnMessagesFromAnotherUser() {
...
+        @Disabled("MAILBOX-384 DeletedMessageVault search will return a wrong result in case of using string " +
+            "equalsIgnoreCase with a special locale")
+        @Test
+        default void shouldReturnsMessageWhenPassingAStringInDifferentLocaleToContainsIgnoreCase() {
+            Locale TURKISH_LOCALE = Locale.forLanguageTag("tr");
+            String subjectInTurkishLanguage = "TITLE";
+            DeletedMessage message1 = storeMessageWithSubject(subjectInTurkishLanguage);
+
+            String subjectLowercase = subjectInTurkishLanguage.toLowerCase(TURKISH_LOCALE);
+            System.out.println(subjectLowercase);
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/DeletedMessageVaultSearchContract.java]: Debug spotted

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged
